enhancements/ingress/logging-headers.md

```diff
@@ -0,0 +1,260 @@
+      destination:
+        type: Container
+      httpCaptureHeaders:
+        request:
+        - name: Host
+          maxLength: 90
+        - name: Referer
+          maxLength: 90
+        response:
+        - name: Content-length
+          maxLength: 9
+        - name: Location
+          maxLength: 90
+      httpLogFormat: '{"time":"%t","clientAddress":"%ci","clientPort":"%cp","frontend":"%f","backend":"%b","server":"%s","request":"%r","response":"%ST","requestLength":"%U","responseLength":"%B","requestHost":"%[capture.req.hdr(0),json(utf8s)]","requestReferer":"%[capture.req.hdr(1),json(utf8s)]","responseContentLength":"%[capture.res.hdr(0),json(utf8s)]","responseLocation":"%[capture.res.hdr(1),json(utf8s)]"}'
+```
+
+This format string produces log messages of the following form:
+
+```
+2020-06-10T17:48:38.013139+00:00 router-default-cd9c9d6c7-xrp5l router-default-cd9c9d6c7-xrp5l haproxy[70]: {"time":"10/Jun/2020:17:48:37.874","clientAddress":"174.19.21.82","clientPort":"33862","frontend":"fe_sni","backend":"be_secure:openshift-console:console","server":"pod:console-64ff858c-ghp2k:console:10.128.0.47:8443","request":"GET /auth/login HTTP/1.1","response":"303","requestLength":"845","responseLength":"975","requestHost":"console-openshift-console.apps.ci-ln-8cjiqnb-d5d6b.origin-ci-int-aws.dev.rhcloud.com","requestReferer":"https:\/\/console-openshift-console.apps.ci-ln-8cjiqnb-d5d6b.origin-ci-int-aws.dev.rhcloud.c","responseContentLength":"341","responseLocation":"https:\/\/oauth-openshift.apps.ci-ln-8cjiqnb-d5d6b.origin-ci-int-aws.dev.rhcloud.com\/oauth\/a"}
+```
+
+Omitting the `spec.logging.access.httpLogFormat` field to use the default format
+string would produce log messages of the following form:
+
+```
+2020-06-10T17:43:59.327608+00:00 router-default-65dccc49cc-2cshx router-default-65dccc49cc-2cshx haproxy[71]: 174.19.21.82:33634 [10/Jun/2020:17:43:59.323] fe_sni~ be_secure:openshift-console:console/pod:console-64ff858c-ghp2k:console:10.128.0.47:8443 0/0/0/3/3 303 975 - - --VN 23/5/0/0/0 0/0 {console-openshift-console.apps.ci-ln-8cjiqnb-d5d6b.origin-ci-int-aws.dev.rhcloud.com|https://console-openshift-console.apps.ci-ln-8cjiqnb-d5d6b.origin-ci-int-aws.dev.rhcloud.c} {341|https://oauth-openshift.apps.ci-ln-8cjiqnb-d5d6b.origin-ci-int-aws.dev.rhcloud.com/oauth/a} "GET /auth/login HTTP/1.1"
+```
+
+#### Validation
+
+Omitting `spec.logging.access.httpCaptureHeaders` or both
+`spec.logging.access.httpCaptureHeaders.request` and
+`spec.logging.access.httpCaptureHeaders.response` specifies the default
+behavior.
+
+The API validates the `spec.logging.access.httpCaptureHeaders.request[*].name`
+and `spec.logging.access.httpCaptureHeaders.response[*].name` field values as
+described by the field type's `+kubebuilder:validation:Pattern` marker.
+Similarly, the API validates that the
+`spec.logging.access.httpCaptureHeaders.request[*].maxLength` and
+`spec.logging.access.httpCaptureHeaders.response[*].maxLength` field values are
+integers.
```

Comment:
Should we cap the `maxLength` to a reasonable value? I question whether problems can be caused to the router (i.e. performance) if `maxLength` is set to an extremely high integer.

Reply:
Maybe... But HAProxy doesn't document a maximum value.  The implementation stores the maximum in an `int`, so we could assume the maximum is 2^31, but that is technically not necessarily true and is also probably already large enough to cause problems.  Any idea how we could arrive at a reasonable maximum?

Reply:
Looks like the default maximum header size accepted by HAProxy is `15360` bytes.
(As per https://github.com/haproxy/haproxy/blob/1385c88560ef5022382ecf683ec933e5b0d45dcc/include/haproxy/defaults.h#L51-L57). 
Would that be a reasonable limit?

Reply:
We set the buffer size to 32768 bytes, and there's a chance it could increase in the future: https://github.com/openshift/router/blob/74299507f283fbbaef7b303df5234a52275f1b6e/images/router/haproxy/conf/haproxy-config.template#L57-L60